File: lc_soln/0424_longest_repeating_character_replacement/solution.cs
public class Solution
{
    public int CharacterReplacement(string s, int k)
    {
        int result = 0;
        int start = 0;
        int high = 0;
        var counter = new Dictionary<char, int>();
        for (int i = 0; i < s.Length; ++i)
        {
            if (!counter.TryAdd(s[i], 1))
            {
                counter[s[i]]++;
            }
            high = int.Max(high, counter[s[i]]);
            while (i - start + 1 - high > k)
            {
                counter[s[start]]--;
                start++;
            }
            result = int.Max(result, i - start + 1);
        }
        return result;
    }
}

File: lc_soln/0424_longest_repeating_character_replacement/solution.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn character_replacement(s: String, k: i32) -> i32 {
        let chars: Vec<char> = s.chars().collect();
        let mut max_len = 0;
        let mut left = 0;
        let mut max_freq = 0;
        let mut freqs = HashMap::new();

        for (right, &c) in chars.iter().enumerate() {
            let count = freqs.entry(c).or_insert(0);
            *count += 1;
            max_freq = max_freq.max(*count);

            // window is invalid when (window_len - max_freq) > k
            while (right - left + 1) as i32 - max_freq > k {
                *freqs.get_mut(&chars[left]).unwrap() -= 1;
                left += 1;
            }

            max_len = max_len.max((right - left + 1) as i32);
        }

        max_len
    }
}

fn main() {}
